### ISSUE TITLE
JSON-file path and unique filenames

### ISSUE THREAD

**Issue by wswebcreation** [cucumber/cucumber-js#926]
After some searching I found the lines of code that write the JSON report, if provided in the `format` option, to a file, see [here](https://github.com/cucumber/cucumber-js/blob/master/src/cli/index.js#L38).

I've got 2 issues / questions about it

1. `fs.open(path.resolve(this.cwd, outputTo), 'w')` is used to create the file. The `w` option _opens the file for writing. The file is created (if it does not exist) or truncated (if it exists)._ But this doesn't create the folder if it's not present and it will fail creating the file if the folder is not present. Is this a conscious choice or a bug? Or are you open for a PR that verifies if the folders already exists?
2. There are cases in which the file name is not unique because the file already exists from a different process / instance. This will overwrite the current report with the data of a new feature.  Is this a conscious choice or are you open for a PR where a unique filename is created when the file already exists?

I think this is related to https://github.com/protractor-cucumber-framework/protractor-cucumber-framework/issues/99

**Comment by charlierudolph** [comment]
1. Creating a the parent folders if they don't exist was not done to keep it simple. I'm fine updating it to create the parent folders if it doesn't exist.

2. I don't like that idea. I think being able to overwrite existing files is important. If you are running multiple processes they should output to different files

**Comment by wswebcreation** [comment]
@charlierudolph

Tnx for your quick reply. I'll create the PR the comming days. A pitty for the second part, I just need to be very inventive to get this working. Still a pitty that the previous listner has been removed.

**Comment by charlierudolph** [comment]
Sorry. The old listener was a quick way to solve a feature request but exposed too much of the internals of cucumber. I'm happy to see it gone and hope we can solve all of its uses in other ways.

**Comment by wswebcreation** [comment]
@charlierudolph 

I think I banged my head to many times against the wall :(. With help of @darrinholst, who checked it with an example project, I found the problem. The problem was sitting between the screen and my chair (bought myself a donkey hat already).

To give you some background. I'm currently writing a protractor plugin that can add instance data to the json report and that had 1 line of code that was cleaning the result folder in the protractor plugin `[setup](https://github.com/angular/protractor/blob/master/lib/plugins.ts#L39)` phase.
According to the documentation the _WebDriver session has been started_ but the _test-framework shouldn't been set up_. It looks like this is wrong because when the WebDriver session has been started Cucumber has already been loaded. This means that Cucumber already did the `fs.open(path.resolve(this.cwd, outputTo), 'w')` meaning that the file is already created. My code removed that folder after the file had been created. The second feature file that created a unique json file with a PID (this is done by @darrinholst his protractor-cucumber-framework module) was pulled up and creating a new file, looking like it was clobbering with the previous one. 

By removing my clean option it worked nicely and I'm able to add data to the report file so I don't need the listener anymore :)

Sorry for bothering you with this and thanks for helping me out with this
